packages/make-styles/src/makeStyles.ts

```diff
@@ -1,42 +1,81 @@
+import {
+  MakeStylesOptions,
+  MakeStylesRenderer,
+  MakeStylesResolvedRule,
+  MakeStylesStyleFunctionRule,
+  MakeStylesStyleRule,
+} from './types';
+
+type Created<Slots extends string> = Record<Slots, Record<string, MakeStylesResolvedRule>>;
+
+function resolveClasses<Slots extends string>(
+  resolvedStyles: Created<Slots>,
+  dir: 'ltr' | 'rtl',
+  renderer: MakeStylesRenderer,
+) {
+  const resolvedClasses = {} as Record<Slots, string>;
+
+  // eslint-disable-next-line guard-for-in
+  for (const slotName in resolvedStyles) {
+    const slotClasses = renderer.insertDefinitions(dir, resolvedStyles[slotName]);
+    const sequenceHash = SEQUENCE_PREFIX + hashString(slotClasses);
+
+    const resultSlotClasses = sequenceHash + ' ' + slotClasses;
+
+    DEFINITION_LOOKUP_TABLE[sequenceHash] = resolvedStyles[slotName];
+    resolvedClasses[slotName] = resultSlotClasses;
+  }
+
+  return resolvedClasses;
+}
 
 export function makeStyles<Slots extends string, Tokens>(
   stylesBySlots: Record<Slots, MakeStylesStyleRule<Tokens>>,
   unstable_cssPriority: number = 0,
 ) {
+  let resolvedStyles: Created<Slots> | null = null;
+
   let resolvedClasses: Record<Slots, string> | null = null;
+  let resolvedClassesRtl: Record<Slots, string> | null = null;
+
   const insertionCache: Record<string, boolean> = {};
 
   function computeClasses(options: MakeStylesOptions<Tokens>): Record<Slots, string> {
-    if (resolvedClasses === null || insertionCache[options.renderer.id] === undefined) {
-      const tokens = createCSSVariablesProxy(options.tokens);
-      resolvedClasses = {} as Record<Slots, string>;
+    const { dir, renderer, tokens } = options;
+
+    if (resolvedStyles === null) {
+      resolvedStyles = {} as Created<Slots>;
+
+      const tokensProxy = createCSSVariablesProxy(tokens);
 
       // eslint-disable-next-line guard-for-in
       for (const slotName in stylesBySlots) {
-        // TODO: Miro says that it should be done once as there is no sense to resolve the same styles
         const slotStyles = stylesBySlots[slotName];
-
         const preparedSlotStyles =
-          typeof slotStyles === 'function' ? (slotStyles as MakeStylesStyleFunctionRule<Tokens>)(tokens) : slotStyles;
-        const resolvedSlotStyles = resolveStyleRules(preparedSlotStyles, unstable_cssPriority);
-
-        const slotClasses = options.renderer.insertDefinitions(resolvedSlotStyles, !!options.rtl);
-        const sequenceHash = SEQUENCE_PREFIX + hashString(slotClasses);
+          typeof slotStyles === 'function'
+            ? (slotStyles as MakeStylesStyleFunctionRule<Tokens>)(tokensProxy)
+            : slotStyles;
 
-        const resultSlotClasses = sequenceHash + ' ' + slotClasses;
-
-        DEFINITION_LOOKUP_TABLE[sequenceHash] = resolvedSlotStyles;
-        resolvedClasses[slotName] = resultSlotClasses;
+        resolvedStyles[slotName] = resolveStyleRules(preparedSlotStyles, unstable_cssPriority);
+      }
+    }
 
+    if (dir === 'rtl') {
```

Comment:
1. have rendererA and rendererB
2. `useStyles('rtl', rendererA)` - OK
3. `useStyles('ltr', rendererB)` - OK
4. `useStyles('rtl', rendererB)` - `resolvedClassesRtl` is not null, `insertionCache[rendererB]` is true -> RTL classes are not inserted into `rendererB` - ERROR

Reply:
Good catch, fixed in e00a213